File: Algo/BST/bst.cpp

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
    ~TreeNode() {delete left, right, val;};
};

TreeNode* search(TreeNode *root, int val) {
    if (!root)
        return nullptr;
    TreeNode* curr=root;
    while (curr) {
        if (curr->val == val)
            return curr;
        if (curr->val < val)
            curr = curr->right;
        else
            curr = curr->left;
    }
    return nullptr;
}

TreeNode* insert(TreeNode *root, int val) {
    if (!root) {
        return new TreeNode(val);
    }
    TreeNode *curr = root;
    while (curr) {
        if (curr->val <= val) {
            if (curr->right == nullptr) {
                curr->right = new TreeNode(val);
                break;
            }
            curr=curr->right;
        } else {
            if (curr->left == nullptr) {
                curr->left = new TreeNode(val);
                break;
            }
            curr=curr->left;
        }
    }
    return root;
}
void go_tail_and_remove(TreeNode *node, TreeNode *par) {
    bool lnode = node->val <= par->val;
    TreeNode *left=node->left, *right=node->right;
    if (left == nullptr) {
        if (lnode) 
            par->left = right;
        else
            par->right = right;
        return;
    } else if (right) {
        TreeNode *tail = right;
        while (tail->left) 
            tail = tail->left;
        tail->left = left->right;
        left->right = right;
    }
    if (lnode)
        par->left = left;
    else
        par->right = right;
}
TreeNode* remove(TreeNode *root, int val) {
    if (!root || (root->val == val && root->left == nullptr))
        return nullptr;
    TreeNode *curr = root, *par = new TreeNode(0, root, nullptr);
    while (curr) {
        if (curr->val == val) {
            go_tail_and_remove(curr, par);
            break;
        }
        if (curr->val < val) {
            par = curr;
            curr = curr->right;
        } else {
            par = curr;
            curr = curr->left;
        }
    }
    return root;
}

File: BST/bst.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def RemoveNode(root: TreeNode, val: int):
    def dfs(par, node):
        if not node:
            return
        if node.val == val:
            delete(par, node)
        elif node.val < val:
            dfs(node, node.right)
        else:
            dfs(node, node.left)
            
    def delete(par, node):
        l, r = node.left, node.right
        lnode = par.val > node.val
        if not l:
            if lnode:
                par.left = r
            else:
                par.right = r
            return
        elif r:
            tail = r
            while tail.left:
                tail = tail.left
            tail.left = l.right
            l.right = r
        if lnode:
            par.left = l
        else:
            par.right = l

    dummy = TreeNode(float('inf'), left = root)
    dfs(dummy, root)
    return dummy.left

def SearchNode(root: TreeNode, val: int):
    if not root:
        return None
    curr = root
    while curr:
        if curr.val == val:
            return curr
        if curr.val < val:
            curr = curr.right
        else:
            curr = curr.left
    return None

def InsertNode(root: TreeNode, val: int):
    if not root:
        return TreeNode(val)
    
    curr = root
    while True:
        if curr.val <= val:
            if not curr.right:
                curr.right = TreeNode(val)
                break
            curr = curr.right
        else:
            if not curr.left:
                curr.left = TreeNode(val)
                break
            curr = curr.left
    return root